### ISSUE TITLE
SmartLocalDefs is often recomputed

### ISSUE THREAD

**Issue by ericbodden** [soot-oss/soot#138]
A simple investigation shows that Soot often recomputes SmartLocalDefs multiple times for the same method (often five times). The analysis is rather costly and should probably be cached instead.

**Comment by badlogic** [comment]
We are hitting this issue as well, and it imposes quite some runtime overhead in our application. Most of the time seems to be spend in the flow analysis according to JProfiler. I saw some recent changes in ForwardFlowAnalysis, did this improve performance?

**Comment by ericbodden** [comment]
Yes, we had some performance improvements in the past year, however the main problem still remains: The conversion to Jimple happens in several passes, each pass of which might change (and often does change) the respective JimpleBody. After each of these passes, the SimpleLocalDefs must be recomputed, unless we find a way to incrementally update it, or to maybe perform some checksums as @patricklam suggested. If I find some time I will look into this a bit.

**Comment by badlogic** [comment]
Great. I ported the latest flow analysis code which had a tiny impact on
the runtime.

**Comment by ericbodden** [comment]
I implemented some pooling in commit 454f7c4
Let's see what our nightly builds say about this. Hope it works. I ran some statistics. If it works then we should safe the creation of about every second SmartLocalDefs instance (and the connected SimpleLiveLocals and ExceptionalUnitGraph instances).

I was unable to update some clients that use DalvikThrowAnalysis. To pool for those, too, we would need to extend the mechanism a bit.

**Comment by badlogic** [comment]
This is awesome! I'll try to backport it to our Soot fork and will report
results. Looks like a simple fix :)

**Comment by ericbodden** [comment]
Note to self: we should find a way to clear the pool at a sensible time, when Jimplification has completed.

**Comment by ericbodden** [comment]
@badlogic You can probably simply cherry-pick the commit via git.
